### PR TITLE
fix: Remove authorization header from core.windows.net requests for Azure docker registries

### LINKED ISSUE #5273: Unable to authorize azurecr.io docker registries
**What Renovate type are you using?**

Self-hosted

**Describe the bug**

Azure docker registries (similarly to Amazon registries) are actually hosted on Azure blob service, which doesn't require an HTTP Authorization be set.  Follows on from #5263 

**To Reproduce**

Try renovating a repo with an azurecr.io.

**Additional context**

After more testing seems there is another issue which I'll submit a PR for.  Relates to https://github.com/renovatebot/renovate/blob/master/lib/datasource/docker/index.ts#L446

Essentially, the Azure registry behaves in a similar way to the Amazon one - it needs to have that authorization header deleted and then it works.

### PR DESCRIPTION
**Pull request by joeapearson**
Closes #5273

### CODE REVIEW

**Comment by joeapearson** [comment]
I think this could really do with some tests - I had a quick look to see how it could be done but I couldn't figure out how to get into that `beforeRedirect` hook without spending more time

**Comment by rarkins** [comment]
> I think this could really do with some tests 

do it live

File: lib/datasource/docker/index.ts [hunk 1]
```diff
@@ -445,6 +445,12 @@ export function getConfigResponse(
             // eslint-disable-next-line no-param-reassign
             delete options.headers.authorization;
```
Comment (inos123) [inline lib/datasource/docker/index.ts]:
```suggestion
            delete options.headers.authorization;
```

**Review by viceice — CHANGES_REQUESTED** [review]
needs unit tests to fix coverage

**Comment by joeapearson** [comment]
Added some tests.  If you prefer not to add an extra export to `docker.ts` then please can you suggest something?  Otherwise I suggest adding an extra file alongside it that exports just this hook function.

File: lib/datasource/docker/index.ts [hunk 2]
```diff
@@ -420,33 +420,41 @@ async function getTags(
...
+    // docker registry is hosted on amazon, redirect url includes authentication.
+    // eslint-disable-next-line no-param-reassign
+    delete options.headers.authorization;
+  }
+
+  if (
+    options.href &&
+    options.href.indexOf('blob.core.windows.net') !== -1 &&
+    options.headers &&
+    options.headers.authorization
```
Comment (ViceIce) [inline lib/datasource/docker/index.ts]:
```suggestion
    options.href?.indexOf('blob.core.windows.net') !== -1 &&
    options.headers?.authorization
```

Simplify condition. Maybe needs prettify.

File: lib/datasource/docker/index.ts [hunk 3]
```diff
@@ -420,33 +420,41 @@ async function getTags(
   }
 }
 
+export function getConfigResponseBeforeRedirectHook(options: any): void {
+  if (options.search && options.search.indexOf('X-Amz-Algorithm') !== -1) {
```
Comment (ViceIce) [inline lib/datasource/docker/index.ts]:
```suggestion
  if (options.search?.indexOf('X-Amz-Algorithm') !== -1) {
```
Simplify condition

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

File: lib/datasource/docker/index.ts [hunk 4]
```diff
@@ -420,33 +420,41 @@ async function getTags(
   }
 }
 
+export function getConfigResponseBeforeRedirectHook(options: any): void {
```
Comment (ViceIce) [inline lib/datasource/docker/index.ts]: We should add a type instead of using `any`.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/docker/index.ts (hunk 3 above)
Comment (rarkins) [reply lib/datasource/docker/index.ts]: btw while we're at it, I prefer the readability if `.includes('foo')` than `.indexOf('foo') !== -1`

File: lib/datasource/docker/index.ts (hunk 4 above)
Comment (ViceIce) [reply lib/datasource/docker/index.ts]: leave it as is, we will do it later. 😅

File: lib/datasource/docker/index.ts (hunk 3 above)
Comment (joeapearson) [reply lib/datasource/docker/index.ts]: voila...

File: lib/datasource/docker/index.ts (hunk 2 above)
Comment (joeapearson) [reply lib/datasource/docker/index.ts]: done

File: lib/datasource/docker/index.ts [hunk 5]
```diff
@@ -420,33 +424,39 @@ async function getTags(
...
+      delete options.port; // Redirect will instead use 80 or 443 for HTTP or HTTPS respectively
+    }
+
+    // docker registry is hosted on amazon, redirect url includes authentication.
+    // eslint-disable-next-line no-param-reassign
+    delete options.headers.authorization;
+  }
+
+  if (
+    options.href?.includes('blob.core.windows.net') &&
```
Comment (ViceIce) [inline lib/datasource/docker/index.ts]:
```suggestion
    options.href.includes('blob.core.windows.net') &&
```

Maybe null check not required, as href is alwys set?

**Review by viceice — APPROVED** [review]
LGTM

File: lib/datasource/docker/index.ts (hunk 5 above)
Comment (rarkins) [reply lib/datasource/docker/index.ts]: But also no major downside, right?

**Pull request merged**